File: prototype/openui-a2ui-cloud-eval/src/a2ui.rs
use std::collections::BTreeMap;

use anyhow::{anyhow, bail, Context};
use serde_json::{Map, Value};

use crate::{ActionSpec, Protocol, ProtocolAdapter, Surface, TypedNode};

pub struct A2UiAdapter;

impl ProtocolAdapter for A2UiAdapter {
    fn normalize(&self, validated_program: &Value, source: &[u8]) -> anyhow::Result<Surface> {
        if validated_program.get("ok") != Some(&Value::Bool(true)) {
            bail!("A2UI oracle did not accept messages");
        }
        let diagnostics = validated_program
            .get("diagnostics")
            .and_then(Value::as_array)
            .context("A2UI oracle missing diagnostics")?;
        if !diagnostics.is_empty() {
            bail!("A2UI diagnostics: {}", serde_json::to_string(diagnostics)?);
        }
        let components = validated_program
            .pointer("/resolved/components")
            .and_then(Value::as_array)
            .context("A2UI oracle missing resolved components")?;

        let mut nodes = BTreeMap::new();
        for component in components {
            let component = component
                .as_object()
                .context("resolved A2UI component must be an object")?;
            let id = string(component, "id")?;
            let kind = string(component, "component")?;
            let props = object(component, "properties")?;
            let node = normalize_component(&id, &kind, props)?;
            if nodes.insert(id.clone(), node).is_some() {
                bail!("duplicate component id: {id}");
            }
        }

        let mut referenced = std::collections::BTreeSet::new();
        for reference in nodes.values().flat_map(TypedNode::references) {
            if !nodes.contains_key(reference) {
                bail!("broken component reference: {reference}");
            }
            referenced.insert(reference);
        }
        let roots = nodes
            .keys()
            .filter(|id| !referenced.contains(id.as_str()))
            .cloned()
            .collect::<Vec<_>>();
        if roots.len() != 1 {
            bail!("A2UI surface must have one root, found {}", roots.len());
        }
        Surface::new(Protocol::A2ui, roots[0].clone(), nodes, source)
    }
}

fn normalize_component(
    id: &str,
    kind: &str,
    props: &Map<String, Value>,
) -> anyhow::Result<TypedNode> {
    let id = id.to_owned();
    Ok(match kind {
        "Text" => TypedNode::Text {
            id,
            text: string(props, "text")?,
        },
        "Stack" => TypedNode::Stack {
            id,
            children: string_array(props, "children")?,
        },
        "Card" => TypedNode::Card {
            id,
            title: string(props, "title")?,
            child: string(props, "child")?,
        },
        "Table" => TypedNode::Table {
            id,
            columns: string_array(props, "columns")?,
            rows: table_rows(props)?,
        },
        "Input" => TypedNode::Input {
            id,
            label: string(props, "label")?,
            state_key: string(props, "state_key")?,
            value: string(props, "value")?,
        },
        "Select" => TypedNode::Select {
            id,
            label: string(props, "label")?,
            state_key: string(props, "state_key")?,
            options: string_array(props, "options")?,
            value: string(props, "value")?,
        },
        "Button" => {
            let action = object(props, "action")?;
            TypedNode::Button {
                id,
                label: string(props, "label")?,
                action: ActionSpec {
                    name: string(action, "name")?,
                    expense_id: string(action, "expense_id")?,
                },
            }
        }
        "Alert" => TypedNode::Alert {
            id,
            tone: string(props, "tone")?,
            message: string(props, "message")?,
        },
        _ => bail!("unknown component: {kind}"),
    })
}

fn object<'a>(value: &'a Map<String, Value>, key: &str) -> anyhow::Result<&'a Map<String, Value>> {
    value
        .get(key)
        .and_then(Value::as_object)
        .with_context(|| format!("missing object: {key}"))
}

fn string(value: &Map<String, Value>, key: &str) -> anyhow::Result<String> {
    value
        .get(key)
        .and_then(Value::as_str)
        .map(str::to_owned)
        .ok_or_else(|| anyhow!("missing string: {key}"))
}

fn string_array(value: &Map<String, Value>, key: &str) -> anyhow::Result<Vec<String>> {
    value
        .get(key)
        .and_then(Value::as_array)
        .with_context(|| format!("missing array: {key}"))?
        .iter()
        .map(|item| {
            item.as_str()
                .map(str::to_owned)
                .ok_or_else(|| anyhow!("{key} must contain strings"))
        })
        .collect()
}

fn table_rows(value: &Map<String, Value>) -> anyhow::Result<Vec<BTreeMap<String, String>>> {
    value
        .get("rows")
        .and_then(Value::as_array)
        .context("Table missing rows")?
        .iter()
        .map(|row| {
            row.as_object()
                .context("table row must be an object")?
                .iter()
                .map(|(key, value)| {
                    value
                        .as_str()
                        .map(|value| (key.clone(), value.to_owned()))
                        .ok_or_else(|| anyhow!("table values must be strings"))
                })
                .collect()
        })
        .collect()
}

File: prototype/openui-a2ui-cloud-eval/oracles/src/a2ui.mjs
import {
  A2uiMessageListWrapperSchema,
  Catalog,
  MessageProcessor,
} from "@a2ui/web_core/v0_9";
import { readFileSync } from "node:fs";
import { z } from "zod3";

const components = [
  { name: "Text", schema: z.object({ text: z.string() }).strict() },
  { name: "Stack", schema: z.object({ children: z.array(z.string()) }).strict() },
  { name: "Card", schema: z.object({ title: z.string(), child: z.string() }).strict() },
  {
    name: "Table",
    schema: z
      .object({ columns: z.array(z.string()), rows: z.array(z.record(z.string())) })
      .strict(),
  },
  {
    name: "Input",
    schema: z
      .object({ label: z.string(), state_key: z.literal("review_note"), value: z.string() })
      .strict(),
  },
  {
    name: "Select",
    schema: z
      .object({
        label: z.string(),
        state_key: z.literal("status_filter"),
        options: z.array(z.string()),
        value: z.string(),
      })
      .strict(),
  },
  {
    name: "Button",
    schema: z
      .object({
        label: z.string(),
        action: z
          .object({ name: z.literal("ApproveExpense"), expense_id: z.string() })
          .strict(),
      })
      .strict(),
  },
  {
    name: "Alert",
    schema: z
      .object({
        tone: z.enum(["info", "success", "warning", "error"]),
        message: z.string(),
      })
      .strict(),
  },
];

export const a2uiCatalog = new Catalog(
  "openui-dioxus-expense-review@0.0.0",
  components,
);

const minimalEnvelope = JSON.parse(
  readFileSync(new URL("../../fixtures/minimal-envelope.a2ui.json", import.meta.url), "utf8"),
);

export function buildA2UiPrompt() {
  const processor = new MessageProcessor([a2uiCatalog], undefined, { version: "v0.9.1" });
  const capabilities = processor.getClientCapabilities({
    includeInlineCatalogs: true,
    version: "v0.9.1",
  });
  return [
    "Return only one A2UI v0.9.1 JSON object with a messages array.",
    "The messages array must contain, in order:",
    "1. A version v0.9.1 createSurface message with surfaceId, catalogId, and sendDataModel true.",
    "2. A version v0.9.1 updateComponents message with the same surfaceId and one components array.",
    "3. A version v0.9.1 updateDataModel message with the same surfaceId, path /, and the state object in value.",
    "Put id, component, and component props directly on each component object.",
    "Put version on every message and nowhere at the top level.",
    "Create exactly one surface with catalogId openui-dioxus-expense-review@0.0.0.",
    "Use every catalog component exactly once and stable ids for all components.",
    "Use only review_note and status_filter as state keys.",
    "Use only ApproveExpense with an expense_id as an action.",
    "Preserve both supplied expense rows exactly.",
    "The Card and Stack child values are component ids. Produce one unreferenced root.",
    "Client capabilities and inline catalog:",
    JSON.stringify(capabilities),
    "Minimal envelope example (syntax only; do not copy its content):",
    JSON.stringify(minimalEnvelope),
  ].join("\n");
}

function cleanJson(source) {
  const trimmed = source.trim();
  const fenced = trimmed.match(/^```(?:json)?\s*([\s\S]*?)\s*```$/i);
  return fenced ? fenced[1] : trimmed;
}

function diagnosticsFromZod(error) {
  return error.issues.map((issue) => ({
    code: issue.code,
    path: issue.path.join("/"),
    message: issue.message,
  }));
}

export function validateA2Ui(source) {
  try {
    const raw = JSON.parse(cleanJson(source));
    const parsed = A2uiMessageListWrapperSchema.safeParse(raw);
    if (!parsed.success) {
      return {
        ok: false,
        protocol: "a2ui",
        version: "v0.9.1/@a2ui/web_core@0.10.6",
        diagnostics: diagnosticsFromZod(parsed.error),
        resolved: null,
      };
    }

    const unknown = parsed.data.messages
      .flatMap((message) => message.updateComponents?.components ?? [])
      .find((component) => !a2uiCatalog.components.has(component.component));
    if (unknown) {
      return {
        ok: false,
        protocol: "a2ui",
        version: "v0.9.1/@a2ui/web_core@0.10.6",
        diagnostics: [
          { code: "unknown-component", message: `Unknown component: ${unknown.component}` },
        ],
        resolved: null,
      };
    }

    const processor = new MessageProcessor([a2uiCatalog], undefined, { version: "v0.9.1" });
    processor.processMessages(parsed.data);
    const surfaces = [...processor.model.surfacesMap.values()];
    if (surfaces.length !== 1) {
      throw new Error(`expected one A2UI surface, found ${surfaces.length}`);
    }
    const surface = surfaces[0];
    const resolved = {
      surface_id: surface.id,
      components: [...surface.componentsModel.entries].map(([id, component]) => ({
        id,
        component: component.type,
        properties: component.properties,
      })),
      data_model: surface.dataModel.get("/"),
    };
    return {
      ok: true,
      protocol: "a2ui",
      version: "v0.9.1/@a2ui/web_core@0.10.6",
      diagnostics: [],
      resolved,
    };
  } catch (error) {
    return {
      ok: false,
      protocol: "a2ui",
      version: "v0.9.1/@a2ui/web_core@0.10.6",
      diagnostics: [{ code: "processor-error", message: String(error?.message ?? error) }],
      resolved: null,
    };
  }
}
